File: cpp/Stack/20.valid-parentheses.cpp
#include "../std.hpp"
class Solution {
public:
    bool isValid(string s)
    {
        stack<char> st;
        for (auto &&c : s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            }
            else {
                if (st.empty()) {
                    return false;
                }
                if (c == ')') {
                    if (st.top() != '(') {
                        return false;
                    }
                    else {
                        st.pop();
                    }
                }
                else if (c == '}') {
                    if (st.top() != '{') {
                        return false;
                    }
                    else {
                        st.pop();
                    }
                }
                else if (c == ']') {
                    if (st.top() != '[') {
                        return false;
                    }
                    else {
                        st.pop();
                    }
                }
            }
        }
        return st.empty();
    }
};
int main()
{
    Solution s;
    s.isValid("]");
}

File: TypeScript/Stack/20.valid-parentheses.ts
function isValid(s: string): boolean {
  const m = {
    "}": "{",
    "]": "[",
    ")": "(",
  }
  let stack: string[] = []
  for (const str of s) {
    if (str in m) {
      if (stack.length != 0 && stack.at(-1) === m[str]) {
        stack.pop()
      } else return false
    } else {
      stack.push(str)
    }
  }
  return stack.length === 0
}
